R/conformal_infer.R

```diff
@@ -0,0 +1,459 @@
+
+  # ------------------------------------------------------------------------------
+  # Get ballpark estimate of range of trial values for new data
+
+  # Use squared residuals as outcome in a gamma model to predict the standard
+  # deviation at a given prediction.
+  var_mod <-
+    try(
+      mgcv::gam(sq ~ s(.pred),
+                data = train_res,
+                family = stats::Gamma(link = "log")),
+      silent = TRUE
+    )
+
+  if (inherits(var_mod, "try-error")) {
+    msg <- c(
+      "!" = "The model to estimate the possible interval length failed with the following message:",
+      "i" = conditionMessage(attr(var_mod, "condition"))
+    )
+    rlang::abort(msg)
+  }
+
+  var_mod
+}
+
+
+setup_new_data <- function(object, new_data, multiplier = 10) {
+  new_pred <- predict(object$wflow, new_data)
+
+  rlang::check_installed("mgcv")
+  var_pred <- as.vector(predict(object$var_mod, new_pred, type = "response"))
+  # convert back to original units
+  var_pred <- sqrt(var_pred)
+  # Add a buffer
+  new_pred$.bound <- multiplier * var_pred
+  new_pred
+
+}
+
+# ------------------------------------------------------------------------------
+# helpers for trial values
+
+# ------------------------------------------------------------------------------
+# For a trial value of the bounds, fit the model and return statistics
+# combine train_data and new_data
+
+trial_fit <- function(trial, trial_data, wflow, level) {
+  y_name <- get_outcome_name(wflow)
+
+  # trial_data is training set with one extra row for the sample being inferred
+  # Last value of the outcome vector is NA; add our trial value there
+  trial_data[[y_name]][nrow(trial_data)] <- trial
+  tmp_fit <- try(fit(wflow, trial_data), silent = TRUE)
+  if (inherits(tmp_fit, "try-error")) {
+    return(
+      dplyr::tibble(quantile = NA_real_, trial = trial, .abs_resid = NA_real_)
+    )
+  }
+
+  # Compute the abs residuals then more statistics
+  tmp_res <-
+    augment(tmp_fit, trial_data) |>
+    dplyr::mutate(
+      .abs_resid = abs(!!rlang::sym(y_name) - .pred)
+    )
+  # Quantiles of original training set (abs) residuals
+  quant_val <- stats::quantile(tmp_res$.abs_resid[-1], prob = level)
+
+  res <-
+    dplyr::tibble(
+      quantile = unname(quant_val),
+      trial = trial,
+      .abs_resid = tmp_res$.abs_resid[nrow(trial_data)]
+    )
+  res$difference <- res$.abs_resid - res$quantile
+  res
+}
+
+# ------------------------------------------------------------------------------
+
+grid_all <- function(new_data, model, train_data, level, ctrl) {
+
+  furrr::future_map_dfr(
+    new_data,
+    grid_one,
+    model,
+    train_data,
+    level,
+    ctrl,
+    .progress = ctrl$progress,
+    .options = furrr::furrr_options(seed = ctrl$seed, stdout = FALSE)
+  )
+}
+
+grid_one <- function(new_data, model, train_data, level, ctrl) {
+  y_name <- get_outcome_name(model)
+
+  pkg_inst <- lapply(ctrl$required_pkgs, rlang::is_installed)
```

Comment:
```suggestion
  pkg_inst <- lapply(ctrl$required_pkgs, rlang::check_installed)
```

`rlang::is_installed` just returns `TRUE` or `FALSE`--I believe we instead want `rlang::check_installed` here.

Reply:
My rationale was that we should stop execution if a package is missing.

Reply:
Okay, great! `check_installed()` will do that for us.